redis_cache.go

```diff
@@ -0,0 +1,124 @@
+package lcw
+
+import (
+	"sync/atomic"
+	"time"
+
+	"github.com/go-redis/redis"
+	"github.com/pkg/errors"
+)
+
+// RedisCache implements LoadingCache for Redis.
+type RedisCache struct {
+	options
+	CacheStat
+	backend *redis.Client
+}
+
+// NewRedisCache makes Redis LoadingCache implementation.
+func NewRedisCache(opts ...Option) (*RedisCache, error) {
+
+	res := RedisCache{
+		options: options{
+			ttl: 5 * time.Minute,
+		},
+	}
+	for _, opt := range opts {
+		if err := opt(&res.options); err != nil {
+			return nil, errors.Wrap(err, "failed to set cache option")
+		}
+	}
+
+	res.backend = redis.NewClient(res.redisOptions)
+
+	return &res, nil
+}
+
+// Get gets value by key or load with fn if not found in cache
+func (c *RedisCache) Get(key string, fn func() (Value, error)) (data Value, err error) {
+
+	v, ok := c.backend.Get(key).Result()
+	if ok == nil {
+		atomic.AddInt64(&c.Hits, 1)
+		return v, nil
+	}
+	if ok == redis.Nil {
+		if data, err = fn(); err != nil {
+			atomic.AddInt64(&c.Errors, 1)
+			return data, err
+		}
+	} else if ok != nil {
+		atomic.AddInt64(&c.Errors, 1)
+		return v, ok
+	}
+	atomic.AddInt64(&c.Misses, 1)
+
+	if c.allowed(key, data) {
+		c.backend.Set(key, data, c.ttl)
+	}
+	return data, nil
+}
+
+// Invalidate removes keys with passed predicate fn, i.e. fn(key) should be true to get evicted
+func (c *RedisCache) Invalidate(fn func(key string) bool) {
+	for _, key := range c.backend.Keys("*").Val() { // Keys() returns copy of cache's key, safe to remove directly
+		if fn(key) {
+			c.backend.Del(key)
+		}
+	}
+}
+
+// Peek returns the key value (or undefined if not found) without updating the "recently used"-ness of the key.
+func (c *RedisCache) Peek(key string) (Value, bool) {
+	ret, err := c.backend.Get(key).Result()
+	if err != nil {
+		return nil, false
+	}
+	return ret, true
+}
+
+// Purge clears the cache completely.
+func (c *RedisCache) Purge() {
```

Comment:
to me it looks like the code here, as well as in over places treats redis as a single-purpose instance. In rela life it can be shared thing used by miltiple consumers

Reply:
maybe it's ok, because I see DB passed in.